### PR TITLE
`ultralytics 8.3.81` Fix Metrics `on_plot` circular references

### LINKED ISSUE #6259: Memory Issue/Leak (RAM piles up infinitely)
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train

### Bug

When training a model with Yolov8, after the training completes, the RAM is not released. 
This becomes an issue when training multiple models iteratively in Python code due to RAM stacking until 99.99%

In my testing I wait for each model to finish training before moving onto the next. I do not use any parallelism. Yet RAM goes to 99.99% by the 4th model training. 

How can we free memory after training completes? 

The current hack I use is to run each model training as its own process and to kill it once training is done. This hack was also suggested by a Ultralytics Community Moderator. But surely this can't be the correct way to handle things?

### Environment

Ultralytics YOLOv8.0.196 🚀 Python-3.9.16 torch-2.0.1+cu117 CUDA:0 (Tesla T4, 15984MiB)

RAM 64GB
Python 3.10
Tested centos, ubuntu, windows

### Minimal Reproducible Example

Train a Yolov8 models.

model.train(task=task, data=data_yaml_path, epochs=5, imgsz=imgsz,
            single_cls=True, cache=False, batch=16, workers=1,
            device=device_list) 

### Additional

Community helper said he suspects Yolo keeps everything in memory until session is closed. But I don't think it is satisfactory to resolve RAM leak issue by restarting the program.

Image after 5 epoch train completes (COCO 128) 
Memory is not freed after completion.
image

Image of that 5 epoch train program on exit
Memory is released
image

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by RemiPT**
Closes https://github.com/ultralytics/ultralytics/issues/6259

`DetectionValidator`, `OBBValidator`, `PoseValidator`, and `SegmentationValidator` all have their own metric object as attribute, which itself is instantiated with their method `on_plot`, which is saved as an attributed of the metric.

This creates a circular reference, and makes it harder to delete both the validator and the metrics.
Unfortunately, the validator can be quite heavy, since it contains `Dataset` and `Dataloader` objects, which, if launched with multiple workers and with `PIN_MEMORY=True`, can seize up quite a lot of CPU memory.

This is not a problem if we only interact with it from the CLI, of from `model.val(...)` once or twice during code execution, but trying to evaluate `model.val(...)` on different checkpoints within the same script will cause an accumulation of CPU memory, and eventually and OOM error.

I wrote a solution to break the reference cycle as little invasively as possible.

It took me great pain to understand where the reference cycle was, but I'm happy to explain it further 🔮.

I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary  
This PR refines the codebase by removing optional callbacks for plot rendering (`on_plot`) from various metric evaluation modules, including detection, segmentation, pose, and oriented bounding boxes (OBBs).

### 📊 Key Changes  
- Removed the `on_plot` parameter from metric initialization (`DetMetrics`, `SegmentMetrics`, `PoseMetrics`, and `OBBMetrics`).
- Updated related methods (`process`) to accept `on_plot` as a direct argument instead of being tied to instance attributes.  
- Minor version update from `8.3.80` to `8.3.81`.

### 🎯 Purpose & Impact  
- **Purpose**: Simplifies the metrics implementation by eliminating unused or unnecessary `on_plot` attributes, streamlining the logic for better readability and maintainability.  
- **Impact**:  
  - **For Developers**: Cleaner code structure for metrics, reducing complexity in metric handling.  
  - **For Users**: Reduced risk of confusion or errors associated with improperly set `on_plot` configurations, leading to more stable and predictable performance.  
  - No functional changes to end-user workflows or output. 🚀

### CODE REVIEW

**Comment by RemiPT** [comment]
I have read the CLA Document and I sign the CLA

**Comment by UltralyticsAssistant** [comment]
👋 Hello @RemiPT, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! Your contribution to improving memory handling and circular reference management is greatly appreciated. To ensure a smooth review process, please consider the following guidelines:

- ✅ **Define a Purpose**: You've done a great job explaining the purpose and significance of your PR. If you have any benchmarks or additional examples demonstrating the CPU memory improvements with your fix, feel free to include them in the description. 
- ✅ **Provide a Minimum Reproducible Example (MRE)**: To validate the issue and your fix, we kindly request an MRE (if not already provided) showing how the OOM issue arises without the patch. This will help confirm the effectiveness of the enhancement.
- ✅ **Synchronize with Source**: Ensure your PR is fully synchronized with the latest `main` branch of the [ultralytics/ultralytics](https://github.com/ultralytics/ultralytics) repository. If it's outdated, please merge the latest changes.
- ✅ **Ensure CI Checks Pass**: Confirm that all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If there are any failures, please resolve them to ensure compatibility.
- ✅ **Update Documentation**: If this change requires modifications to the usage or involves new utilities (like `detach_circular_reference`), consider adding or updating the [documentation](https://docs.ultralytics.com/) accordingly.
- ✅ **Add Tests**: If applicable, include or update unit tests to verify the behavior introduced (e.g., testing `detach_circular_reference`). Ensure all tests pass to validate your changes.
- ✅ **Sign the CLA**: Thank you for signing the [Contributor License Agreement (CLA)](https://docs.ultralytics.com/help/CLA/). This ensures compliance with our contribution policy!

For more detailed instructions, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Let us know if you have any questions, as we're here to assist.

This is an automated response 🤖, but rest assured, an Ultralytics engineer will review your PR soon. Thank you again for your valuable contribution! 🚀👏

**Comment by rpiche-flyscan** [comment]
I have read the CLA Document and I sign the CLA

**Comment by Burhan-Q** [comment]
It might be good to include a before/after speed and memory comparison when running validation. I suspect that if benefit can be shown that it's reducing one of the two, it would improve the likelihood of the PR getting accepted. Either way, _I_ appreciate this, because it's a concept I hadn't considered and an area I'll have to read about more!

**Comment by glenn-jocher** [comment]
Thank you for your thoughtful suggestion and engagement with the PR! While we agree comparative metrics can be valuable, this particular fix primarily addresses memory reference cycles that accumulate during repeated `model.val()` calls in long-running processes rather than immediate speed gains. For contributors, we recommend including CPU memory profiling comparisons using tools like Python's `tracemalloc` or `memory-profiler` when submitting similar optimizations. Would you like guidance on implementing memory benchmarking for this case? 

This PR review will follow our standard technical evaluation process once fully synced with the latest `main` branch and passing all CI checks.

**Comment by RemiPT** [comment]
Hi Glenn, Burhan,
I'll be happy to include benchmarks to complement this PR, and would greatly appreciate any resource you could point me to for performing CPU memory profiling comparisons.

Thank you for your response,
Remi

**Comment by Burhan-Q** [comment]
You can use the `utils.py:benchmark` to check performance speeds. Regarding memory usage, you can use `psutil.virtual_memory()` to measure used/available memory usage.

**Comment by glenn-jocher** [comment]
Thank you for your contribution, Remi! For CPU memory profiling, we recommend using Python's built-in `tracemalloc` module alongside our existing `torch_utils.profile` function shown in the Ultralytics Torch Utilities documentation. Here's a minimal example to track memory allocations during validation:

```python
import tracemalloc
from ultralytics import YOLO

tracemalloc.start()
model = YOLO("yolo11n.pt")
model.val()  # Run validation
current, peak = tracemalloc.get_traced_memory()
print(f"Peak memory usage: {peak / 1e6}MB")
```

This approach will help quantify memory improvements from your circular reference fix. Please include these benchmarks in your PR when ready! 🚀

**Comment by glenn-jocher** [comment]
Thank you for the detailed memory analysis and reproducible test case. The benchmark data clearly demonstrates your circular reference fix effectively mitigates memory accumulation during repeated validations, especially in scenarios like multi-checkpoint evaluation. This strengthens the PR's value for users running batch validations or hyperparameter tuning.

We'll prioritize technical review once the PR is fully synced with `main` and passes all CI checks. Please ensure synchronization when ready, and we appreciate your thorough approach to memory optimization in the validation pipeline.

**Comment by Laughing-q** [comment]
@RemiPT Hey thanks for the PR! yeah I'm able to reproduce the same memory increasing issue here when applying validation many times in the same script.
@glenn-jocher @Burhan-Q @RemiPT I think the simplest way here to do is to directly remove the attribute `on_plot` in all the `Metrics` object since we only need it in `Metrics.process` method and we could directly pass the argument to the function, no need to assign it as an attribute here.

**Comment by RemiPT** [comment]
@Laughing-q Looks good! Tested the memory usage, looks like it's working on my end as well

**Review by Laughing-q — APPROVED** [review]

**Comment by UltralyticsAssistant** [comment]
🎉 Big news: The PR by @RemiPT with amazing contributions from @Laughing-q, @rpiche-flyscan, @Burhan-Q, and @glenn-jocher has been successfully merged! 🛠️✨  

By refining the metric handling across detection, segmentation, pose, and OBBs, you've not only streamlined our code but made it smarter, cleaner, and more maintainable for everyone. These changes, though seemingly subtle, ripple outward to improve the experience for developers and users alike. Your innovation matters! 🚀  

As Leonardo da Vinci once said, *"Simplicity is the ultimate sophistication."* By simplifying what matters, you've taken us one step closer to excellence. Thank you all for your exceptional collaboration and thoughtful contributions—your work is shaping the future! 🌟

**Comment by glenn-jocher** [comment]
Thanks to everyone involved in identifying and resolving this circular reference issue! Memory management is critical for production environments, especially when working with large datasets or running repeated validations. This fix elegantly addresses a subtle but impactful problem by restructuring how the `on_plot` functionality is handled in our metrics implementation.

The memory benchmarks clearly demonstrate the value of breaking these reference cycles. This improvement will be particularly beneficial for users running validation on multiple checkpoints or datasets within a single process, as well as for longer-running applications.

Great teamwork all around - these kinds of optimizations significantly improve the reliability of the Ultralytics framework.
